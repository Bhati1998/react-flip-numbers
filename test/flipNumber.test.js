import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import FlipNumber from './flipNumber';

jest.mock('react-simple-animate', () => 'Animate');

const numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

describe('FlipNumber', () => {
  const props = {
    height: 20,
    color: 'white',
    background: 'black',
    numbers,
    width: 20,
    perspective: '1000px',
    durationSeconds: 0.3,
    activeNumber: 0,
    delaySeconds: 0,
    startAnimation: true,
  };

  it('should render correctly', () => {
    const tree = renderer.create(<FlipNumber {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should only update component when active number changed or degree is 0', () => {
    const tree = shallow(<FlipNumber {...props} />);

    const nextProps = {
      ...props,
      activeNumber: 0,
    };

    expect(tree.instance().shouldComponentUpdate(nextProps)).toBeTruthy();

    tree.setState({
      degree: 1,
      isStatic: false,
    });

    const nextState = {
      degree: 1,
      rotateCounter: 0,
      rotateDegreePerNumber: 0,
      isStatic: false,
    };

    expect(tree.instance().shouldComponentUpdate(nextProps, nextState)).toBeFalsy();

    expect(
      tree.instance().shouldComponentUpdate({
        activeNumber: 1,
      })
    ).toBeTruthy();
  });

  it('should return the correct position for matching number', () => {
    const tree = renderer.create(
      <FlipNumber
        {...{
          ...props,
          activeNumber: 2,
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it('should reset the rotate position when it is been running for 1000 times', () => {
    const tree = shallow(<FlipNumber {...props} />);

    tree.setState({
      rotateCounter: 1001,
    });

    tree.setProps({
      numbers,
      activeNumber: 0,
    });

    expect(tree.state()).toMatchObject({
      rotateCounter: 0,
    });
  });
});
