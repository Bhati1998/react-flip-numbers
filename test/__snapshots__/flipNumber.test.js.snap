// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipNumber should render correctly 1`] = `
<span
  aria-hidden={true}
  style={
    Object {
      "display": "inline-block",
      "height": 20,
      "overflow": "hidden",
      "perspective": "1000px",
      "textAlign": "left",
      "width": "20px",
    }
  }
>
  <Animate
    delaySeconds={0}
    durationSeconds={0.3}
    easeType="cubic-bezier(0.19, 1, 0.22, 1)"
    endStyle={
      Object {
        "height": "100%",
        "position": "absolute",
        "transform": "rotateX(0deg)",
        "transformStyle": "preserve-3d",
      }
    }
    startStyle={
      Object {
        "height": "100%",
        "position": "absolute",
        "transformStyle": "preserve-3d",
      }
    }
    tag="span"
  >
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(0deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      0
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(36deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(72deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(108deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(144deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      4
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(180deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      5
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(216deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      6
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(252deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      7
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(288deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      8
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(324deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      9
    </span>
  </Animate>
  <span
    data={NaN}
    style={
      Object {
        "WebkitFontSmoothing": "antialiased",
        "alignItems": "center",
        "background": "black",
        "color": "white",
        "display": "flex",
        "fontSize": "19px",
        "height": 20,
        "justifyContent": "center",
        "left": "0px",
        "lineHeight": "20px",
        "position": "absolute",
        "textAlign": "center",
        "transform": "rotateX(0deg) translateZ(30px)",
        "visibility": "hidden",
        "width": "20px",
      }
    }
  >
    0
  </span>
</span>
`;

exports[`FlipNumber should return the correct position for matching number 1`] = `
<span
  aria-hidden={true}
  style={
    Object {
      "display": "inline-block",
      "height": 20,
      "overflow": "hidden",
      "perspective": "1000px",
      "textAlign": "left",
      "width": "20px",
    }
  }
>
  <Animate
    delaySeconds={0}
    durationSeconds={0.3}
    easeType="cubic-bezier(0.19, 1, 0.22, 1)"
    endStyle={
      Object {
        "height": "100%",
        "position": "absolute",
        "transform": "rotateX(-72deg)",
        "transformStyle": "preserve-3d",
      }
    }
    startStyle={
      Object {
        "height": "100%",
        "position": "absolute",
        "transformStyle": "preserve-3d",
      }
    }
    tag="span"
  >
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(0deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      0
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(36deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(72deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(108deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(144deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      4
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(180deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      5
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(216deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      6
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(252deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      7
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(288deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      8
    </span>
    <span
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "alignItems": "center",
          "background": "black",
          "color": "white",
          "display": "flex",
          "fontSize": "19px",
          "height": 20,
          "justifyContent": "center",
          "lineHeight": "20px",
          "position": "absolute",
          "textAlign": "center",
          "transform": "rotateX(324deg) translateZ(30px)",
          "width": "20px",
        }
      }
    >
      9
    </span>
  </Animate>
  <span
    data={NaN}
    style={
      Object {
        "WebkitFontSmoothing": "antialiased",
        "alignItems": "center",
        "background": "black",
        "color": "white",
        "display": "flex",
        "fontSize": "19px",
        "height": 20,
        "justifyContent": "center",
        "left": "0px",
        "lineHeight": "20px",
        "position": "absolute",
        "textAlign": "center",
        "transform": "rotateX(0deg) translateZ(30px)",
        "visibility": "hidden",
        "width": "20px",
      }
    }
  >
    2
  </span>
</span>
`;
